src/plugins/kibana_utils/common/persistable_state/index.ts

```diff
@@ -65,16 +68,36 @@ export interface PersistableState<P extends SerializableState = SerializableStat
    * @param state
    * @param version
    */
-  migrate?: (state: SerializableState, version: string) => SerializableState;
+  migrate: (state: SerializableState, version: string) => SerializableState;
 }
 
-export type PersistableStateDefinition<P extends SerializableState = SerializableState> = Partial<
-  Omit<PersistableState<P>, 'migrate'>
-> & {
+export interface PersistableState<P extends SerializableState = SerializableState> {
+  /**
+   *  function to extract telemetry information
+   * @param state
+   * @param collector
+   */
+  telemetry: (state: P, collector: Record<string, any>) => Record<string, any>;
+  /**
+   * inject function receives state and a list of references and should return state with references injected
+   * default is identity function
+   * @param state
+   * @param references
+   */
+  inject: (state: P, references: SavedObjectReference[]) => P;
+  /**
+   * extract function receives state and should return state with references extracted and array of references
+   * default returns same state with empty reference array
+   * @param state
+   */
+  extract: (state: P) => { state: P; references: SavedObjectReference[] };
+
   /**
    * list of all migrations per semver
    */
-  migrations?: {
-    [key: string]: MigrateFunction;
-  };
-};
+  migrations: MigrateFunctionsObject;
+}
+
+export type PersistableStateDefinition<P extends SerializableState = SerializableState> = Partial<
+  PersistableState<P>
+>;
```

Comment:
Could you describe the changes here just a little bit more? I'm not 100% sure when you would want to use which type.

It looks like `PersistableStateService` and `PersistableState` are identical other than `migrate` vs `migrations`. 

`PersistableStateDefinition` is now just a shortcut way of calling `Partial<PersistableState>`.

What would be the use case for when to use each? If all 3 are necessary maybe we can add some comments to clear this up.

Reply:
PersistableStateDefinition - is for definitions (embeddable factory, expression function etc), where things are mostly optional
PersistableState - actual items  in the registry, where things are no longer optional but have defaults provided
PersistableStateService - service exposing a registry with persistable state items

yes, the main difference is that on definition thigns are optional, and the service exposes migrate function instead of migrations object

Reply:
Thanks for clarifying, makes sense.

My only minor nit then would be considering doing a shared interface for telemetry/extract/inject since they are the same, and then using the shared interface in `PersistableState` and `PersistableStateService`. But I don't feel too strongly on that; since they are already in the same file it should be pretty easy to remember to update it in both places if we ever need to